File: jebat_sdk/src/jebat_sdk/retry.py
"""
JEBAT SDK Retry Logic

Retry configuration with exponential backoff using tenacity.
"""

from typing import Callable, Type, TypeVar, Union
from tenacity import (
    retry,
    stop_after_attempt,
    wait_exponential,
    retry_if_exception_type,
    retry_if_exception,
    before_sleep_log,
    after_log,
)
import logging

logger = logging.getLogger(__name__)

T = TypeVar("T")

# Default retry configuration
DEFAULT_MAX_ATTEMPTS = 3
DEFAULT_MIN_WAIT = 1  # seconds
DEFAULT_MAX_WAIT = 10  # seconds
DEFAULT_MULTIPLIER = 2

# Exception types that should trigger a retry
RETRYABLE_EXCEPTIONS = (
    ConnectionError,
    TimeoutError,
    IOError,
)


def is_retryable_exception(exc: BaseException) -> bool:
    """Determine if an exception should trigger a retry."""
    from .exceptions import (
        ServerError,
        TimeoutError,
        ConnectionError,
        RateLimitError,
    )

    # Retry on network/connection errors
    if isinstance(exc, RETRYABLE_EXCEPTIONS):
        return True

    # Retry on server errors (5xx)
    if isinstance(exc, ServerError):
        return True

    # Retry on rate limit with retry-after
    if isinstance(exc, RateLimitError):
        return True

    # Don't retry on client errors (4xx except 429)
    return False


def get_retry_decorator(
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
    min_wait: float = DEFAULT_MIN_WAIT,
    max_wait: float = DEFAULT_MAX_WAIT,
    multiplier: float = DEFAULT_MULTIPLIER,
    retryable_exceptions: tuple = RETRYABLE_EXCEPTIONS,
) -> Callable:
    """
    Create a retry decorator with exponential backoff.

    Args:
        max_attempts: Maximum number of attempts
        min_wait: Minimum wait time in seconds
        max_wait: Maximum wait time in seconds
        multiplier: Exponential multiplier
        retryable_exceptions: Tuple of exception types to retry

    Returns:
        A retry decorator
    """
    return retry(
        reraise=True,
        stop=stop_after_attempt(max_attempts),
        wait=wait_exponential(
            multiplier=multiplier,
            min=min_wait,
            max=max_wait,
        ),
        retry=retry_if_exception_type(retryable_exceptions) | retry_if_exception(is_retryable_exception),
        before_sleep=before_sleep_log(logger, logging.WARNING),
        after=after_log(logger, logging.INFO),
    )


def retry_with_backoff(
    func: Callable[..., T],
    *args,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
    min_wait: float = DEFAULT_MIN_WAIT,
    max_wait: float = DEFAULT_MAX_WAIT,
    multiplier: float = DEFAULT_MULTIPLIER,
    retryable_exceptions: tuple = RETRYABLE_EXCEPTIONS,
    **kwargs,
) -> T:
    """
    Execute a function with exponential backoff retry.

    Args:
        func: Function to execute
        max_attempts: Maximum number of attempts
        min_wait: Minimum wait time
        max_wait: Maximum wait time
        multiplier: Exponential multiplier
        retryable_exceptions: Exception types to retry
        *args, **kwargs: Arguments to pass to func

    Returns:
        Function result
    """
    retry_decorator = get_retry_decorator(
        max_attempts=max_attempts,
        min_wait=min_wait,
        max_wait=max_wait,
        multiplier=multiplier,
        retryable_exceptions=retryable_exceptions,
    )
    return retry_decorator(func)(*args, **kwargs)


async def async_retry_with_backoff(
    func: Callable[..., T],
    *args,
    max_attempts: int = DEFAULT_MAX_ATTEMPTS,
    min_wait: float = DEFAULT_MIN_WAIT,
    max_wait: float = DEFAULT_MAX_WAIT,
    multiplier: float = DEFAULT_MULTIPLIER,
    retryable_exceptions: tuple = RETRYABLE_EXCEPTIONS,
    **kwargs,
) -> T:
    """
    Execute an async function with exponential backoff retry.

    Args:
        func: Async function to execute
        max_attempts: Maximum number of attempts
        min_wait: Minimum wait time
        max_wait: Maximum wait time
        multiplier: Exponential multiplier
        retryable_exceptions: Exception types to retry
        *args, **kwargs: Arguments to pass to func

    Returns:
        Function result
    """
    retry_decorator = get_retry_decorator(
        max_attempts=max_attempts,
        min_wait=min_wait,
        max_wait=max_wait,
        multiplier=multiplier,
        retryable_exceptions=retryable_exceptions,
    )
    return await retry_decorator(func)(*args, **kwargs)


# Pre-configured retry decorators for common use cases

# Quick retry for transient errors (2 retries, fast)
quick_retry = get_retry_decorator(
    max_attempts=2,
    min_wait=0.5,
    max_wait=2,
    multiplier=1,
)

# Standard retry for API calls (3 retries, moderate)
standard_retry = get_retry_decorator(
    max_attempts=3,
    min_wait=1,
    max_wait=10,
    multiplier=2,
)

# Aggressive retry for critical operations (5 retries, slow)
aggressive_retry = get_retry_decorator(
    max_attempts=5,
    min_wait=2,
    max_wait=30,
    multiplier=2,
)

# Rate limit aware retry (respects retry-after header)
rate_limit_retry = get_retry_decorator(
    max_attempts=5,
    min_wait=1,
    max_wait=60,
    multiplier=2,
    retryable_exceptions=(Exception,),  # We'll filter in is_retryable_exception
)

File: jebat-sdk-ts/src/retry.ts
/**
 * Retry logic with exponential backoff for JEBAT SDK.
 */

import type { RetryConfig } from './types';
import { isRetryableError, JebatError } from './exceptions';

/**
 * Default retry configuration.
 */
export const DEFAULT_RETRY_CONFIG: Required<RetryConfig> = {
  maxAttempts: 3,
  minWait: 1000,   // 1 second
  maxWait: 10000,  // 10 seconds
  multiplier: 2,
};

/**
 * Calculate wait time with exponential backoff and jitter.
 */
export function calculateWaitTime(
  attempt: number,
  config: Required<RetryConfig>
): number {
  const exponentialWait = config.minWait * Math.pow(config.multiplier, attempt);
  const cappedWait = Math.min(exponentialWait, config.maxWait);
  // Add jitter (±25%)
  const jitter = cappedWait * 0.25 * (Math.random() * 2 - 1);
  return Math.max(config.minWait, Math.round(cappedWait + jitter));
}

/**
 * Sleep utility.
 */
export function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Retry decorator for async functions.
 */
export function withRetry<T extends (...args: unknown[]) => Promise<unknown>>(
  fn: T,
  config: RetryConfig = {}
): T {
  const mergedConfig = { ...DEFAULT_RETRY_CONFIG, ...config };

  return (async (...args: unknown[]) => {
    let lastError: unknown;

    for (let attempt = 0; attempt < mergedConfig.maxAttempts; attempt++) {
      try {
        return await fn(...args);
      } catch (error) {
        lastError = error;

        // Don't retry on last attempt
        if (attempt === mergedConfig.maxAttempts - 1) {
          break;
        }

        // Check if error is retryable
        if (!isRetryableError(error)) {
          throw error;
        }

        // Special handling for rate limit with retry-after
        if (error instanceof JebatError && 'retryAfter' in error) {
          const rateLimitError = error as { retryAfter: number | null };
          if (rateLimitError.retryAfter) {
            await sleep(rateLimitError.retryAfter * 1000);
            continue;
          }
        }

        // Wait before retry
        const waitTime = calculateWaitTime(attempt, mergedConfig);
        await sleep(waitTime);
      }
    }

    throw lastError;
  }) as T;
}

/**
 * Async retry helper function.
 */
export async function retry<T>(
  fn: () => Promise<T>,
  config: RetryConfig = {}
): Promise<T> {
  const mergedConfig = { ...DEFAULT_RETRY_CONFIG, ...config };
  let lastError: unknown;

  for (let attempt = 0; attempt < mergedConfig.maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;

      if (attempt === mergedConfig.maxAttempts - 1) {
        break;
      }

      if (!isRetryableError(error)) {
        throw error;
      }

      if (error instanceof JebatError && 'retryAfter' in error) {
        const rateLimitError = error as { retryAfter: number | null };
        if (rateLimitError.retryAfter) {
          await sleep(rateLimitError.retryAfter * 1000);
          continue;
        }
      }

      const waitTime = calculateWaitTime(attempt, mergedConfig);
      await sleep(waitTime);
    }
  }

  throw lastError;
}

/**
 * Create a retryable fetch wrapper.
 */
export function createRetryableFetch(config: RetryConfig = {}) {
  return async (url: string | URL, init?: RequestInit): Promise<Response> => {
    return retry(async () => {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), (init as any)?.signal ? 0 : 30000);

      try {
        const response = await fetch(url, {
          ...init,
          signal: controller.signal,
        });

        // Don't throw on HTTP errors, let caller handle
        return response;
      } finally {
        clearTimeout(timeoutId);
      }
    }, config);
  };
}

/**
 * Pre-configured retry configs for common use cases.
 */
export const RetryPresets = {
  /** Quick retry for transient errors (2 attempts, fast) */
  quick: { maxAttempts: 2, minWait: 500, maxWait: 2000, multiplier: 1 },

  /** Standard retry for API calls (3 attempts, moderate) */
  standard: { maxAttempts: 3, minWait: 1000, maxWait: 10000, multiplier: 2 },

  /** Aggressive retry for critical operations (5 attempts, slow) */
  aggressive: { maxAttempts: 5, minWait: 2000, maxWait: 30000, multiplier: 2 },

  /** Rate limit aware retry (respects retry-after header) */
  rateLimitAware: { maxAttempts: 5, minWait: 1000, maxWait: 60000, multiplier: 2 },
} as const satisfies Record<string, RetryConfig>;

// Re-export RetryConfig from types for convenience
export type { RetryConfig } from './types';